## Gathering Data

In [1]:
#Import all packages needed
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline
import requests
import tweepy 
import json

####  The WeRateDogs Twitter archive

In [2]:
#load the WeRateDogs Twitter archive
df_arch = pd.read_csv('twitter-archive-enhanced.csv')

In [3]:
df_arch.head()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   
1  892177421306343426                    NaN                  NaN   
2  891815181378084864                    NaN                  NaN   
3  891689557279858688                    NaN                  NaN   
4  891327558926688256                    NaN                  NaN   

                   timestamp  \
0  2017-08-01 16:23:56 +0000   
1  2017-08-01 00:17:27 +0000   
2  2017-07-31 00:18:03 +0000   
3  2017-07-30 15:58:51 +0000   
4  2017-07-29 16:00:24 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   
1  <a href="http://twitter.com/download/iphone" r...   
2  <a href="http://twitter.com/download/iphone" r...   
3  <a href="http://twitter.com/download/iphone" r...   
4  <a href="http://twitter.com/download/iphone" r...   

                                                text  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                  NaN   
1  This is Tilly. She's just checking pup on you....                  NaN   
2  This is Archie. He is a rare Norwegian Pouncin...                  NaN   
3  This is Darla. She commenced a snooze mid meal...                  NaN   
4  This is Franklin. He would like you to stop ca...                  NaN   

   retweeted_status_user_id retweeted_status_timestamp  \
0                       NaN                        NaN   
1                       NaN                        NaN   
2                       NaN                        NaN   
3                       NaN                        NaN   
4                       NaN                        NaN   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   
1  https://twitter.com/dog_rates/status/892177421...                13   
2  https://twitter.com/dog_rates/status/891815181...                12   
3  https://twitter.com/dog_rates/status/891689557...                13   
4  https://twitter.com/dog_rates/status/891327558...                12   

   rating_denominator      name doggo floofer pupper puppo  
0                  10   Phineas  None    None   None  None  
1                  10     Tilly  None    None   None  None  
2                  10    Archie  None    None   None  None  
3                  10     Darla  None    None   None  None  
4                  10  Franklin  None    None   None  None

#### The tweet image predictions

In [4]:
#load The tweet image predictions
url = "https://d17h27t6h515a5.cloudfront.net/topher/2017/August/599fd2ad_image-predictions/image-predictions.tsv"
response = requests.get(url)

with open('image-predictions.tsv', mode ='wb') as file:
    file.write(response.content)

df_image = pd.read_csv('image-predictions.tsv', sep='\t' )
df_image.head()

tweet_id                                          jpg_url  \
0  666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1  666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2  666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3  666044226329800704  https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4  666049248165822465  https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   

   img_num                      p1   p1_conf  p1_dog                  p2  \
0        1  Welsh_springer_spaniel  0.465074    True              collie   
1        1                 redbone  0.506826    True  miniature_pinscher   
2        1         German_shepherd  0.596461    True            malinois   
3        1     Rhodesian_ridgeback  0.408143    True             redbone   
4        1      miniature_pinscher  0.560311    True          Rottweiler   

    p2_conf  p2_dog                   p3   p3_conf  p3_dog  
0  0.156665    True    Shetland_sheepdog  0.061428    True  
1  0.074192    True  Rhodesian_ridgeback  0.072010    True  
2  0.138584    True           bloodhound  0.116197    True  
3  0.360687    True   miniature_pinscher  0.222752    True  
4  0.243682    True             Doberman  0.154629    True

In [5]:
df_image = pd.read_csv('image-predictions.tsv', sep='\t' )
df_image.head()

tweet_id                                          jpg_url  \
0  666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1  666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2  666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3  666044226329800704  https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4  666049248165822465  https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   

   img_num                      p1   p1_conf  p1_dog                  p2  \
0        1  Welsh_springer_spaniel  0.465074    True              collie   
1        1                 redbone  0.506826    True  miniature_pinscher   
2        1         German_shepherd  0.596461    True            malinois   
3        1     Rhodesian_ridgeback  0.408143    True             redbone   
4        1      miniature_pinscher  0.560311    True          Rottweiler   

    p2_conf  p2_dog                   p3   p3_conf  p3_dog  
0  0.156665    True    Shetland_sheepdog  0.061428    True  
1  0.074192    True  Rhodesian_ridgeback  0.072010    True  
2  0.138584    True           bloodhound  0.116197    True  
3  0.360687    True   miniature_pinscher  0.222752    True  
4  0.243682    True             Doberman  0.154629    True

#### Retweet count and favorite count Data

In [6]:
# Downloading data using Twitter API
# twitter api
consumer_key = 'wNF635o1tVSsDF7bwLUdQR56x'
consumer_secret = 'sznJmA8FkTv0llA65xMpESkJsRVJmJZ6V1HTC4IAVXZ986Ds1U'
access_token = '2433451075-z2YLaUTkVlzxs9cVuHWbDc8AzNUT4I9t2E8pM9Y'
access_secret = 'nYGmeWwhzOR7sR5DQAU5Phk83VI0FMj3nu5mjDzoLJTh3'

auth = tweepy.OAuthHandler(consumer_key, consumer_secret)
auth.set_access_token(access_token, access_secret)

api = tweepy.API(auth, wait_on_rate_limit = True, wait_on_rate_limit_notify = True)

In [7]:

tweet_ids = df_arch.tweet_id.values
with open('tweet_json.txt', 'a', encoding='utf8') as f:
    for tweet_id in tweet_ids:
        try:
            tweet = api.get_status(tweet_id, tweet_mode='extended')
            json.dump(tweet._json, f)
            f.write('\n')
        except:
            continue

Rate limit reached. Sleeping for: 738
Rate limit reached. Sleeping for: 739


In [8]:
# Converting relevant JSON data to dataframe

tweets_list =[]

with open('tweet_json.txt') as json_file:
    for line in json_file:
    
        tweets_dict = {}
        tweets_json = json.loads(line)
        
        try:
            tweets_dict['tweet_id'] = tweets_json['extended_entities']['media'][0]['id']
        except:
            tweets_dict['tweet_id'] = 'na'

        tweets_dict['retweet_count'] = tweets_json['retweet_count']
        tweets_dict['favorite_count'] = tweets_json['favorite_count']
        
        tweets_list.append(tweets_dict)

In [9]:
df_tweets = pd.DataFrame(tweets_list)
df_tweets.head()

favorite_count  retweet_count            tweet_id
0           39467           8853  892420639486877696
1           33819           6514  892177413194625024
2           25461           4328  891815175371796480
3           42908           8964  891689552724799489
4           41048           9774  891327551943041024

## Assessing Data

####  The WeRateDogs Twitter archive

In [10]:
df_arch.sample(10)

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
2331  666353288456101888                    NaN                  NaN   
1078  739485634323156992                    NaN                  NaN   
1633  684241637099323392                    NaN                  NaN   
1573  687494652870668288                    NaN                  NaN   
1656  683357973142474752                    NaN                  NaN   
2184  668988183816871936                    NaN                  NaN   
605   798576900688019456                    NaN                  NaN   
833   768554158521745409                    NaN                  NaN   
2294  667138269671505920                    NaN                  NaN   
994   748568946752774144                    NaN                  NaN   

                      timestamp  \
2331  2015-11-16 20:32:58 +0000   
1078  2016-06-05 15:54:48 +0000   
1633  2016-01-05 05:14:53 +0000   
1573  2016-01-14 04:41:12 +0000   
1656  2016-01-02 18:43:31 +0000   
2184  2015-11-24 03:03:06 +0000   
605   2016-11-15 17:22:24 +0000   
833   2016-08-24 21:02:45 +0000   
2294  2015-11-19 00:32:12 +0000   
994   2016-06-30 17:28:39 +0000   

                                                 source  \
2331  <a href="http://twitter.com/download/iphone" r...   
1078  <a href="http://twitter.com/download/iphone" r...   
1633  <a href="http://twitter.com/download/iphone" r...   
1573  <a href="http://twitter.com/download/iphone" r...   
1656  <a href="http://twitter.com/download/iphone" r...   
2184  <a href="http://twitter.com/download/iphone" r...   
605   <a href="http://twitter.com/download/iphone" r...   
833   <a href="http://twitter.com/download/iphone" r...   
2294  <a href="http://twitter.com/download/iphone" r...   
994   <a href="http://twitter.com/download/iphone" r...   

                                                   text  retweeted_status_id  \
2331  Here we have a mixed Asiago from the Galápagos...                  NaN   
1078  This is Kyle. He's a heavy drinker and an avid...                  NaN   
1633  This is Obi. He got camera shy. 12/10 https://...                  NaN   
1573  This is Marq. He stole this car. 7/10 wtf Marq...                  NaN   
1656  "Have a seat, son. There are some things we ne...                  NaN   
2184  Honor to rate this dog. Lots of fur on him. Tw...                  NaN   
605   RT @dog_rates: Not familiar with this breed. N...         6.661041e+17   
833   RT @dog_rates: This is Nollie. She's waving at...         7.399792e+17   
2294  Extremely intelligent dog here. Has learned to...                  NaN   
994   This is Cora. She rings a bell for treats. 12/...                  NaN   

      retweeted_status_user_id retweeted_status_timestamp  \
2331                       NaN                        NaN   
1078                       NaN                        NaN   
1633                       NaN                        NaN   
1573                       NaN                        NaN   
1656                       NaN                        NaN   
2184                       NaN                        NaN   
605               4.196984e+09  2015-11-16 04:02:55 +0000   
833               4.196984e+09  2016-06-07 00:36:02 +0000   
2294                       NaN                        NaN   
994                        NaN                        NaN   

                                          expanded_urls  rating_numerator  \
2331  https://twitter.com/dog_rates/status/666353288...                 8   
1078  https://twitter.com/dog_rates/status/739485634...                 6   
1633  https://twitter.com/dog_rates/status/684241637...                12   
1573  https://twitter.com/dog_rates/status/687494652...                 7   
1656  https://twitter.com/dog_rates/status/683357973...                10   
2184  https://twitter.com/dog_rates/status/668988183...                 7   
605   https://twitter.com/dog_rates/status/666104133...                 1   
833   https://twit

In [11]:
df_arch.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
tweet_id                      2356 non-null int64
in_reply_to_status_id         78 non-null float64
in_reply_to_user_id           78 non-null float64
timestamp                     2356 non-null object
source                        2356 non-null object
text                          2356 non-null object
retweeted_status_id           181 non-null float64
retweeted_status_user_id      181 non-null float64
retweeted_status_timestamp    181 non-null object
expanded_urls                 2297 non-null object
rating_numerator              2356 non-null int64
rating_denominator            2356 non-null int64
name                          2356 non-null object
doggo                         2356 non-null object
floofer                       2356 non-null object
pupper                        2356 non-null object
puppo                         2356 non-null object
dtypes: float64(4), int64(3), ob

In [12]:
df_arch.isnull().sum()

tweet_id                         0
in_reply_to_status_id         2278
in_reply_to_user_id           2278
timestamp                        0
source                           0
text                             0
retweeted_status_id           2175
retweeted_status_user_id      2175
retweeted_status_timestamp    2175
expanded_urls                   59
rating_numerator                 0
rating_denominator               0
name                             0
doggo                            0
floofer                          0
pupper                           0
puppo                            0
dtype: int64

In [13]:
df_arch.duplicated().sum()

0

In [14]:
df_arch.rating_numerator.value_counts()

12      558
11      464
10      461
13      351
9       158
8       102
7        55
14       54
5        37
6        32
3        19
4        17
1         9
2         9
420       2
0         2
15        2
75        2
80        1
20        1
24        1
26        1
44        1
50        1
60        1
165       1
84        1
88        1
144       1
182       1
143       1
666       1
960       1
1776      1
17        1
27        1
45        1
99        1
121       1
204       1
Name: rating_numerator, dtype: int64

In [15]:
df_arch.rating_denominator.value_counts()

10     2333
11        3
50        3
80        2
20        2
2         1
16        1
40        1
70        1
15        1
90        1
110       1
120       1
130       1
150       1
170       1
7         1
0         1
Name: rating_denominator, dtype: int64

In [16]:
df_arch.name.value_counts()

None            745
a                55
Charlie          12
Oliver           11
Cooper           11
Lucy             11
Lola             10
Penny            10
Tucker           10
Winston           9
Bo                9
Sadie             8
the               8
Toby              7
Bailey            7
an                7
Buddy             7
Daisy             7
Bella             6
Scout             6
Jax               6
Milo              6
Dave              6
Stanley           6
Leo               6
Rusty             6
Oscar             6
Jack              6
Koda              6
Chester           5
               ... 
Napolean          1
Barry             1
Jeffri            1
Bronte            1
Cilantro          1
Pip               1
Walker            1
Bonaparte         1
Reagan            1
Tug               1
his               1
Eazy              1
Arya              1
Beckham           1
Bodie             1
Maude             1
unacceptable      1
Bilbo             1
Severus           1


In [17]:
df_arch.doggo.value_counts()

None     2259
doggo      97
Name: doggo, dtype: int64

In [18]:
df_arch.floofer.value_counts()

None       2346
floofer      10
Name: floofer, dtype: int64

In [19]:
df_arch.pupper.value_counts()

None      2099
pupper     257
Name: pupper, dtype: int64

In [20]:
df_arch.puppo.value_counts()

None     2326
puppo      30
Name: puppo, dtype: int64

#### The tweet image predictions

In [21]:
df_image.sample(10)

tweet_id                                            jpg_url  \
1911  853639147608842240    https://pbs.twimg.com/media/C9i8RhhXoAAdkMT.jpg   
1019  710140971284037632    https://pbs.twimg.com/media/Cdrtcr-W4AAqi5H.jpg   
451   674743008475090944    https://pbs.twimg.com/media/CV0rL7RWEAAbhqm.jpg   
232   670417414769758208    https://pbs.twimg.com/media/CU3NE8EWUAEVdPD.jpg   
2037  884562892145688576    https://pbs.twimg.com/media/DEaZQkfXUAEC7qB.jpg   
1667  812781120811126785    https://pbs.twimg.com/media/C0eUHfWUAAANEYr.jpg   
979   707297311098011648    https://pbs.twimg.com/media/CdDTJLMW4AEST--.jpg   
959   705591895322394625    https://pbs.twimg.com/media/CcrEFQdUcAA7CJf.jpg   
1924  857263160327368704    https://pbs.twimg.com/media/C-WcS4MXoAADrBU.jpg   
1708  817827839487737858  https://pbs.twimg.com/ext_tw_video_thumb/81782...   

      img_num                    p1   p1_conf  p1_dog                 p2  \
1911        1       German_shepherd  0.509879    True           malinois   
1019        1              Pekinese  0.953170    True           papillon   
451         1  Bernese_mountain_dog  0.583054    True  Shetland_sheepdog   
232         1            sea_urchin  0.493257   False          porcupine   
2037        1                   pug  0.546406    True     French_bulldog   
1667        1          bull_mastiff  0.989316    True              boxer   
979         1      Blenheim_spaniel  0.370717    True           Shih-Tzu   
959         1               basenji  0.877207    True  Italian_greyhound   
1924        1               Samoyed  0.998021    True         Pomeranian   
1708        1        cocker_spaniel  0.387608    True   golden_retriever   

       p2_conf  p2_dog                          p3   p3_conf  p3_dog  
1911  0.237311    True                      kelpie  0.046916    True  
1019  0.019517    True            Japanese_spaniel  0.005821    True  
451   0.065990    True  Greater_Swiss_Mountain_dog  0.065236    True  
232   0.460565   False                     cardoon  0.008146   False  
2037  0.404291    True           Brabancon_griffon  0.044002    True  
1667  0.007043    True              French_bulldog  0.001740    True  
979   0.201566    True         black-footed_ferret  0.101559   False  
959   0.047854    True          miniature_pinscher  0.035638    True  
1924  0.000922    True                    keeshond  0.000311    True  
1708  0.264844    True                    Pekinese  0.122123    True

In [22]:
df_image.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 12 columns):
tweet_id    2075 non-null int64
jpg_url     2075 non-null object
img_num     2075 non-null int64
p1          2075 non-null object
p1_conf     2075 non-null float64
p1_dog      2075 non-null bool
p2          2075 non-null object
p2_conf     2075 non-null float64
p2_dog      2075 non-null bool
p3          2075 non-null object
p3_conf     2075 non-null float64
p3_dog      2075 non-null bool
dtypes: bool(3), float64(3), int64(2), object(4)
memory usage: 152.1+ KB


In [23]:
df_image.isnull().sum()

tweet_id    0
jpg_url     0
img_num     0
p1          0
p1_conf     0
p1_dog      0
p2          0
p2_conf     0
p2_dog      0
p3          0
p3_conf     0
p3_dog      0
dtype: int64

In [24]:
df_image.duplicated().sum()

0

In [25]:
df_image.p1_dog.value_counts()

True     1532
False     543
Name: p1_dog, dtype: int64

In [26]:
df_image.p1.nunique()

378

#### Retweet count and favorite count Data

In [27]:
df_tweets.sample(10)

favorite_count  retweet_count            tweet_id
2768           12746           2299  821044517277233152
3490            4112           1523  724004596214038529
2930               0           6748                  na
72                 0           1349                  na
1480            2622            925                  na
3350            9513           3878                  na
3047            9072           2566  784517508296478720
202            16508           3928  853299949064388608
2136             801            118  670040288568610816
53             51522          11099  881666587375304705

In [28]:
df_tweets.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4685 entries, 0 to 4684
Data columns (total 3 columns):
favorite_count    4685 non-null int64
retweet_count     4685 non-null int64
tweet_id          4685 non-null object
dtypes: int64(2), object(1)
memory usage: 109.9+ KB


In [29]:
df_tweets.isnull().sum()

favorite_count    0
retweet_count     0
tweet_id          0
dtype: int64

In [30]:
df_tweets.tweet_id.duplicated().sum()

2676

In [31]:
df_tweets.tweet_id.value_counts()

na                    553
819004789207928832      4
771770449999097856      4
671896801591607296      4
793286466235408384      4
733109473259085826      4
782969083092471809      4
691416861947379712      4
679828437910089729      4
822489041455157248      4
819006388206649344      4
788765906553962498      4
815965888126062592      4
789530855911882752      4
667509363477577728      4
826958645422342144      4
777684227185868800      4
801167887901724672      4
796149732779421699      4
750429289032642560      4
712809014606827520      4
679158360810352640      4
790277108719386624      4
790946047744966656      4
741067297519869955      4
673320125483892736      4
773547591439122432      4
791672307924209664      4
783334603142598656      4
701214695424991232      4
                     ... 
682429474584018946      2
667793397835542529      2
673662511204319236      2
689154306549792769      2
677228864888823810      2
760289324994879489      2
674038229604044801      2
687460500490

In [32]:
df_tweets[df_tweets.tweet_id == 777684227185868800]

favorite_count  retweet_count            tweet_id
560                0           3403  777684227185868800
765            12518           3403  777684227185868800
2898               0           2958  777684227185868800
3101           11297           2958  777684227185868800

### Quality

#### The WeRateDogs Twitter archive
    - there are 181 retweets and 78 reply tweets(only want original ratings)
    - rating_denominator have invalid values (should be 10)
    - rating_rating_numerator have invalid values 
    - Incorrect names (a,None,...)
    - timestamp column is a string format 
    - extra columns
#### The tweet image predictions
    - missing data
    - extra columns
    - change p1,.. to  prediction1,..

#### Retweet count and favorite count Data
    - duplicated tweet ids and some na ids
    - tweet_id is object 
    - missing some tweets

### Tidiness

- join the three tabels
- df_arch dog stage columns should be merged


## Cleaning Data

In [33]:
#make copy to clean 
df_arch_clean = df_arch.copy()
df_image_clean = df_image.copy()
df_tweets_clean = df_tweets.copy()

####  The WeRateDogs Twitter archive

##### Define

**remove all retweets and replys**

##### Code

In [34]:
df_arch_clean = df_arch_clean.query('retweeted_status_id == "NaN"')

In [35]:
df_arch_clean = df_arch_clean.query('in_reply_to_user_id == "NaN"')

##### Test

In [36]:
df_arch_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2097 entries, 0 to 2355
Data columns (total 17 columns):
tweet_id                      2097 non-null int64
in_reply_to_status_id         0 non-null float64
in_reply_to_user_id           0 non-null float64
timestamp                     2097 non-null object
source                        2097 non-null object
text                          2097 non-null object
retweeted_status_id           0 non-null float64
retweeted_status_user_id      0 non-null float64
retweeted_status_timestamp    0 non-null object
expanded_urls                 2094 non-null object
rating_numerator              2097 non-null int64
rating_denominator            2097 non-null int64
name                          2097 non-null object
doggo                         2097 non-null object
floofer                       2097 non-null object
pupper                        2097 non-null object
puppo                         2097 non-null object
dtypes: float64(4), int64(3), object(10)

##### Define

**change timestamp column type to datetime**

##### Code

In [37]:
df_arch_clean['timestamp'] = pd.to_datetime(df_arch_clean['timestamp'])

##### Test

In [38]:
df_arch_clean.dtypes

tweet_id                               int64
in_reply_to_status_id                float64
in_reply_to_user_id                  float64
timestamp                     datetime64[ns]
source                                object
text                                  object
retweeted_status_id                  float64
retweeted_status_user_id             float64
retweeted_status_timestamp            object
expanded_urls                         object
rating_numerator                       int64
rating_denominator                     int64
name                                  object
doggo                                 object
floofer                               object
pupper                                object
puppo                                 object
dtype: object

##### Define

**remove rating_denominator invalid values**

##### Code

In [39]:
pd.options.display.max_colwidth = 200
df_arch_clean[df_arch_clean.rating_denominator != 10].iloc[:,np.r_[0,5,10,11]]

tweet_id  \
433   820690176645140481   
516   810984652412424192   
902   758467244762497024   
1068  740373189193256964   
1120  731156023742988288   
1165  722974582966214656   
1202  716439118184652801   
1228  713900603437621249   
1254  710658690886586372   
1274  709198395643068416   
1351  704054845121142784   
1433  697463031882764288   
1635  684222868335505415   
1662  682962037429899265   
1779  677716515794329600   
1843  675853064436391936   
2335  666287406224695296   

                                                                                                                                                                           text  \
433                                                                         The floofs have been released I repeat the floofs have been released. 84/70 https://t.co/NIYC820tmd   
516   Meet Sam. She smiles 24/7 &amp; secretly aspires to be a reindeer. \nKeep Sam smiling by clicking and sharing this link:\nhttps://t.co/98tB8y7y7t https://t.co/LouL5vdvxx   
902                                                                                              Why does this never happen at my front door... 165/150 https://t.co/HmwrdfEfUE   
1068                               After so many requests, this is Bretagne. She was the last surviving 9/11 search dog, and our second ever 14/10. RIP https://t.co/XAVDNDaVgQ   
1120                                                  Say hello to this unbelievably well behaved squad of doggos. 204/170 would try to pet all at once https://t.co/yGQI3He3xv   
1165                                                                                                           Happy 4/20 from the squad! 13/10 for all https://t.co/eV1diwds8a   
1202                                                      This is Bluebert. He just saw that both #FinalFur match ups are split 50/50. Amazed af. 11/10 https://t.co/Kky1DPG4iq   
1228                                                                              Happy Saturday here's 9 puppers on a bench. 99/90 good work everybody https://t.co/mpvaVxKmc1   
1254                                                               Here's a brigade of puppers. All look very prepared for whatever happens next. 80/80 https://t.co/0eb7R1Om12   
1274                         From left to right:\nCletus, Jerome, Alejandro, Burp, &amp; Titson\nNone know where camera is. 45/50 would hug all at once https://t.co/sedre1ivTK   
1351                                                                                         Here is a whole flock of puppers.  60/50 I'll take the lot https://t.co/9dpcw6MdWa   
1433                                                                               Happy Wednesday here's a bucket of pups. 44/40 would pet all at once https://t.co/HppvrYuamZ   
1635                               Someone help the girl is being mugged. Several are distracting her while two steal her shoes. Clever puppers 121/110 https://t.co/1zfnTJLt55   
1662                                This is Darrel. He just robbed a 7/11 and is in a high speed police chase. Was just spotted by the helicopter 10/10 https://t.co/7EsP8LmSp5   
1779                                                                                                     IT'S PUPPERGEDDON. Total of 144/120 ...I think https://t.co/ZanVtAtvIq   
1843                                                                Here we have an entire platoon of puppers. Total score: 88/80 would pet all at once https://t.co/y93p6FLvVw   
2335                                   This is an Albanian 3 1/2 legged  Episcopalian. Loves well-polished hardwood flooring. Penis on the collar. 9/10 https://t.co/d9NcXFKwLv   

      rating_numerator  rating_denominator  
433                 84                  70  
516                 24                   7  
902                165                 150  
1068                 9                  11  
1120               204                 170  
1165              

from what we see in the text column we can chane some rating_denominator values manually:
- 722974582966214656 4/20 to 13/10 
- 666287406224695296 1/2 to 9/10
- 682962037429899265 7/11 to 10/10
- 716439118184652801 50/50 to 11/10
- 740373189193256964 9/11 to 14/10
then remove other rows

In [40]:
#manual changes

mask = df_arch_clean.tweet_id == 740373189193256964
df_arch_clean.loc[mask, 'rating_numerator'] = 14
df_arch_clean.loc[mask, 'rating_denominator'] = 10

mask = df_arch_clean.tweet_id == 722974582966214656
df_arch_clean.loc[mask, 'rating_numerator'] = 13
df_arch_clean.loc[mask, 'rating_denominator'] = 10

mask = df_arch_clean.tweet_id == 666287406224695296
df_arch_clean.loc[mask, 'rating_numerator'] = 9
df_arch_clean.loc[mask, 'rating_denominator'] = 10

mask = df_arch_clean.tweet_id == 682962037429899265
df_arch_clean.loc[mask, 'rating_numerator'] = 10
df_arch_clean.loc[mask, 'rating_denominator'] = 10

mask = df_arch_clean.tweet_id == 716439118184652801
df_arch_clean.loc[mask, 'rating_numerator'] = 11
df_arch_clean.loc[mask, 'rating_denominator'] = 10

In [41]:
#drop all rows where denominator not 10
df_arch_clean= df_arch_clean[df_arch_clean.rating_denominator == 10]

##### Test

In [42]:
df_arch_clean[df_arch_clean.rating_denominator != 10].iloc[:,np.r_[0,5,10,11]]

Empty DataFrame
Columns: [tweet_id, text, rating_numerator, rating_denominator]
Index: []

##### Define

**remove rating_numerator invalid values**

##### Code

In [43]:
df_arch_clean.rating_numerator.value_counts()

12      486
10      437
11      414
13      288
9       153
8        98
7        51
14       39
5        34
6        32
3        19
4        15
2         9
1         4
75        1
420       1
26        1
27        1
1776      1
0         1
Name: rating_numerator, dtype: int64

In [44]:
df_arch_clean = df_arch_clean[df_arch_clean.rating_numerator < 15]

##### Test

In [45]:
df_arch_clean.rating_numerator.value_counts()

12    486
10    437
11    414
13    288
9     153
8      98
7      51
14     39
5      34
6      32
3      19
4      15
2       9
1       4
0       1
Name: rating_numerator, dtype: int64

##### Define

**change invalid names**

##### Code

In [46]:
df_arch_clean.name = df_arch_clean.name.str.replace('^[a-z]+', 'None')

##### Test

In [47]:
df_arch_clean.name.value_counts()

None           694
Lucy            11
Charlie         11
Cooper          10
Oliver          10
Penny            9
Tucker           9
Winston          8
Lola             8
Sadie            8
Toby             7
Daisy            7
Bo               6
Jax              6
Bella            6
Koda             6
Bailey           6
Oscar            6
Stanley          6
Chester          5
Rusty            5
Buddy            5
Scout            5
Dave             5
Bentley          5
Leo              5
Louis            5
Milo             5
Jeffrey          4
Boomer           4
              ... 
Stormy           1
Alf              1
Maya             1
Zeek             1
Socks            1
DayZ             1
Cecil            1
Eleanor          1
Chase            1
Pumpkin          1
Sprout           1
Rey              1
William          1
Shakespeare      1
Skittles         1
Tove             1
Pablo            1
Bruiser          1
Fletcher         1
Tycho            1
Schnozz          1
Hubertson   

##### Define

**remove unrequired columns**

##### Code

In [48]:
df_arch_clean = df_arch_clean.drop(['in_reply_to_status_id','in_reply_to_user_id','source','text',
                                    'retweeted_status_id','retweeted_status_user_id','retweeted_status_timestamp',
                                   'expanded_urls','rating_denominator'],axis = 1)                  

##### Test

In [49]:
list(df_arch_clean)

['tweet_id',
 'timestamp',
 'rating_numerator',
 'name',
 'doggo',
 'floofer',
 'pupper',
 'puppo']

In [50]:
df_arch_clean.shape[0]

2080

##### Define

**create stage column**

##### Code

In [51]:

df_arch_clean.loc[df_arch_clean['doggo'] == 'doggo', 'stage'] = 'doggo'
df_arch_clean.loc[df_arch_clean['floofer'] == 'floofer', 'stage'] = 'floofer'
df_arch_clean.loc[df_arch_clean['pupper'] == 'pupper', 'stage'] = 'pupper'
df_arch_clean.loc[df_arch_clean['puppo'] == 'puppo', 'stage'] = 'puppo'
df_arch_clean.drop(['doggo','floofer','pupper','puppo'],axis=1,inplace =True)

##### Test

In [52]:
df_arch_clean.sample(10)

tweet_id           timestamp  rating_numerator      name  \
1990  672640509974827008 2015-12-04 04:56:09                 9  Leonidas   
505   812781120811126785 2016-12-24 22:04:54                10   Swagger   
1041  743980027717509120 2016-06-18 01:33:55                11      Geno   
573   801167903437357056 2016-11-22 20:58:07                13      Milo   
1719  680161097740095489 2015-12-24 23:00:17                12   Goliath   
1355  703611486317502464 2016-02-27 16:03:45                10   Scooter   
1918  674271431610523648 2015-12-08 16:56:51                10      None   
1967  673342308415348736 2015-12-06 03:24:51                10    Django   
1475  693642232151285760 2016-01-31 03:49:30                 9  Freddery   
1331  705591895322394625 2016-03-04 03:13:11                11      None   

       stage  
1990     NaN  
505      NaN  
1041     NaN  
573      NaN  
1719     NaN  
1355  pupper  
1918     NaN  
1967  pupper  
1475     NaN  
1331     NaN

In [53]:
df_arch_clean.stage.value_counts()

pupper     229
doggo       72
puppo       24
floofer     10
Name: stage, dtype: int64

#### The tweet image predictions

##### Define

**rename columns**

##### Code

In [54]:
#in this analysis we well use prediction1 one only
df_image_clean=df_image_clean.rename(columns ={'p1':'prediction','p1_conf':'prediction_conf','p1_dog':'prediction_dog'})

###### Test

In [55]:
list(df_image_clean)

['tweet_id',
 'jpg_url',
 'img_num',
 'prediction',
 'prediction_conf',
 'prediction_dog',
 'p2',
 'p2_conf',
 'p2_dog',
 'p3',
 'p3_conf',
 'p3_dog']

##### Define

**drop extra columns**

##### Code

In [56]:
df_image_clean = df_image_clean.drop(['jpg_url','img_num','p2','p2_conf','p2_dog','p3','p3_conf','p3_dog'],axis = 1)

###### Test

In [57]:
df_image_clean.sample(5)

tweet_id          prediction  prediction_conf  prediction_dog
1974  869772420881756160            Pembroke         0.980148            True
41    666701168228331520  Labrador_retriever         0.887707            True
1321  756303284449767430    golden_retriever         0.981652            True
638   681281657291280384       Saint_Bernard         0.998830            True
164   668975677807423489              basset         0.605437            True

#### Retweet count and favorite count Data

##### Define

**remove rows where is a "na" value and where is a duplicated id**

##### Code

In [58]:
#drop rows where id = na
df_tweets_clean = df_tweets_clean.query('tweet_id != "na"')

In [59]:
df_tweets_clean.tweet_id.duplicated().sum()

2124

In [60]:
#drop rows with duplicated ids
df_tweets_clean = df_tweets_clean.drop_duplicates(subset='tweet_id',keep='last')


###### Test

In [61]:
df_tweets_clean.tweet_id.duplicated().sum()

0

In [62]:
df_tweets_clean.query('tweet_id == "na"')

Empty DataFrame
Columns: [favorite_count, retweet_count, tweet_id]
Index: []

##### Define

**chane tweet_id to int type**

##### Code

In [63]:
df_tweets_clean['tweet_id'] = df_tweets_clean['tweet_id'].astype(int)

##### Test

In [64]:
#datatype for tweet_id in all dataframes

print(df_image_clean['tweet_id'].dtypes)
print(df_arch_clean['tweet_id'].dtypes)
print(df_tweets_clean['tweet_id'].dtypes)

int64
int64
int64


In [65]:
print(df_image_clean[df_image_clean.tweet_id.isnull()].count()['tweet_id'])
print(df_arch_clean[df_arch_clean.tweet_id.isnull()].count()['tweet_id'])
print(df_tweets_clean[df_tweets_clean.tweet_id.isnull()].count()['tweet_id'])

0
0
0


##### Define

**merge all dataframes**

##### Code

In [66]:
tweeter_df = pd.merge(df_arch_clean, df_image_clean, how='inner', on=['tweet_id'])

In [67]:
tweeter_df.shape[0]

1954

In [68]:
tweeter_df = pd.merge(tweeter_df, df_tweets_clean, right_on='tweet_id', left_on='tweet_id', how='inner')

##### Test

In [69]:

tweeter_df.shape[0]

0

In [70]:
tweeter_df.sample(20)

ValueError: a must be greater than 0

##### Define

**tweet that have favorite and retweet count only**

##### Code

In [ ]:
tweeter_df = tweeter_df.query(' prediction != "NaN" & favorite_count != "NaN"')

##### Test

In [ ]:
tweeter_df.head()

## Analyzing and Visualizing Data 

####  The WeRateDogs Twitter archive

#### The tweet image predictions

#### Retweet count and favorite count Data